core/controller/src/main/scala/whisk/core/controller/actions/PrimitiveActions.scala

```diff
@@ -140,6 +200,349 @@ protected[actions] trait PrimitiveActions {
+      case Some(action) if action.annotations.isTruthy(WhiskActivation.conductorAnnotation) => // composition
+        // invokeComposition will increase the invocation counts
+        invokeComposition(
+          user,
+          action,
+          payload,
+          waitForResponse = None, // not topmost, hence blocking, no need for timeout
+          cause = Some(session.activationId),
+          accounting = Some(session.accounting))
+      case Some(action) => // primitive action
+        session.accounting.components += 1
+        invokeSimpleAction(
+          user,
+          action,
+          payload,
+          waitForResponse = Some(action.limits.timeout.duration + 1.minute),
+          cause = Some(session.activationId))
+      case None => // sequence
+        session.accounting.components += 1
+        val SequenceExecMetaData(components) = action.exec
+        invokeSequence(
+          user,
+          action,
+          components,
+          payload,
+          waitForOutermostResponse = None,
+          cause = Some(session.activationId),
+          topmost = false,
+          atomicActionsCount = 0).map(r => r._1)
+    }
+
+    waitForActivation(user, session, activationResponse).flatMap {
+      case Left(response) => // unsuccessful invocation, return error response
+        Future.successful(response)
+      case Right(activation) => // reinvoke conductor on component result
+        invokeConductor(user, payload = Some(activation.resultAsJson), session = session)
+    }
+  }
+
+  /**
+   * Waits for a response from a conductor of component action invocation.
+   * Handles internal errors (activation failure or timeout).
+   * Logs the activation id and updates the duration and max memory for the session.
+   * Returns the activation record if successful, the error response if not.
+   *
+   * @param user the identity invoking the action
+   * @param session the session object for this composition
+   * @param activationResponse the future activation to wait on
+   * @param transid a transaction id for logging
+   */
+  private def waitForActivation(user: Identity,
+                                session: Session,
+                                activationResponse: Future[Either[ActivationId, WhiskActivation]])(
+    implicit transid: TransactionId): Future[Either[ActivationResponse, WhiskActivation]] = {
+
+    activationResponse
+      .map {
+        case Left(activationId) => // invocation timeout
+          session.logs += activationId
+          Left(ActivationResponse.whiskError(compositionActivationTimeout(activationId)))
+        case Right(activation) => // successful invocation
+          session.logs += activation.activationId
+          // activation.duration should be defined but this is not reflected by the type so be defensive
+          session.duration += activation.duration.getOrElse(activation.end.toEpochMilli - activation.start.toEpochMilli)
```

Comment:
Maybe go for `end` - `start` straight away to avoid confusion?

Reply:
I am not sure I understand the suggestion here: `duration` is not equal to `end - start` and we have to use the correct value. I'll update the comment to be more clear.